### PR TITLE
[X86][CodeGen] Prefer KMOVkk_EVEX than KMOVkk when EGPR is supported

### PR DESCRIPTION
**Pull request by KanRobert**
In memory fold table, we have

```
 {X86::KMOVDkk, X86::KMOVDkm, 0},
 {X86::KMOVDkk_EVEX, X86::KMOVDkm_EVEX, 0}
```

where `KMOVDkm_EVEX` can use EGPR as base and index registers, while
`KMOVDkm` can't. Hence, though `KMOVkk` does not have any GPR operands,
we prefer to use `KMOVDkk_EVEX` to help register allocation.

It will be compressed to `KMOVDkk` in EVEX2VEX pass if memory folding
does not happen.

### CODE REVIEW

File: llvm/lib/Target/X86/X86DomainReassignment.cpp [hunk 1]
```diff
@@ -625,8 +625,10 @@ void X86DomainReassignment::initConverters() {
   createReplacerDstCOPY(X86::MOVZX64rm16,
                         HasEGPR ? X86::KMOVWkm_EVEX : X86::KMOVWkm);
 
-  createReplacerDstCOPY(X86::MOVZX32rr16, X86::KMOVWkk);
-  createReplacerDstCOPY(X86::MOVZX64rr16, X86::KMOVWkk);
+  createReplacerDstCOPY(X86::MOVZX32rr16,
+                        HasEGPR ? X86::KMOVWkk_EVEX : X86::KMOVWkk);
+  createReplacerDstCOPY(X86::MOVZX64rr16,
+                        HasEGPR ? X86::KMOVWkk_EVEX : X86::KMOVWkk);
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86DomainReassignment.cpp:631]: Can we access + repurpose the EVEX->VEX tables to do this for us?

Comment (KanRobert) [reply llvm/lib/Target/X86/X86DomainReassignment.cpp:631]: Are you suggesting the usage like
```
  createReplacerDstCOPY(X86::MOVZX64rr16,
                        getEVEXVariant(X86::KMOVWkk, /*UseEVEXVariant=*/HasEGPR));
```

Comment (KanRobert) [reply llvm/lib/Target/X86/X86DomainReassignment.cpp:631]: I think it's probably not worth it. First, KMOV may be the only instruction that needs to be promoted in CPP, and other instructions are implemented in TD through predicate HasEGPR/NoEGPR. Second, looking up the table will change the complexity from O(1) to O(lgN) and increase compile time.

File: llvm/test/CodeGen/X86/apx/kmov-kk.ll [hunk 2]
```diff
@@ -0,0 +1,14 @@
+; RUN: llc < %s -mtriple=x86_64-unknown -mattr=+avx512f,+egpr -show-mc-encoding | FileCheck --check-prefix=EGPR %s
+
+define <16 x i32> @kmovkk(ptr %base, <16 x i32> %ind, i16 %mask) {
+; EGPR: kmovq   %k1, %k2                        # EVEX TO VEX Compression encoding: [0xc4,0xe1,0xf8,0x90,0xd1]
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/apx/kmov-kk.ll:4]: Is it possible to add a test to cover the memory fold case?

Comment (KanRobert) [reply llvm/test/CodeGen/X86/apx/kmov-kk.ll:4]: It's not straightforward to me.  And I think the memory fold is already tested by ./llvm/test/TableGen/x86-fold-tables.inc

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**